<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>ABBA by SystemsGeneticsSG</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">ABBA</h1>
      <h2 class="project-tagline">Approximate Bayes for Bisulfite sequencing Analysis</h2>
      <a href="https://github.com/SystemsGeneticsSG/ABBA" class="btn">View on GitHub</a>
      <a href="https://github.com/SystemsGeneticsSG/ABBA/zipball/master" class="btn">Download .zip</a>
      <a href="example_results.html" class="btn">View Example</a>
    </section>

    <section class="main-content">
      <h1>
<a id="abba" class="anchor" href="#abba" aria-hidden="true"><span class="octicon octicon-link"></span></a>ABBA</h1>

<p>Aproximate Bayes for Bisulphite sequecing Analysis is perl program for analysis WGBS data. The program can be run locally or spread over a cluster using the qsub system. </p>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>About</h3>

<p>
ABBA is a bayesian framework for the identification of differential methylation, taking the data from read counts to regions.
</p>
<p>
<img src="images/ifitm3.png" alt="IFITM3" style="width:100%;">
</p>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h3>

<p>To use ABBA you will are required to have Perl installed and the ability to install packages using CPAN or CPANM. In the situation where you do not have sudo access we recommend that you install PerlBrew and use CPANM to manage your perl libraries, for details see <a href="http://perlbrew.pl">http://perlbrew.pl</a> or follow the following instructions. Please be aware that this may not be suitable for all situations, it is best to try with the system perl first and use this as a fall back.</p>

<h4>
<a id="step-1-install-perlbrew-optional-only-if-your-system-perl-is-not-usable" class="anchor" href="#step-1-install-perlbrew-optional-only-if-your-system-perl-is-not-usable" aria-hidden="true"><span class="octicon octicon-link"></span></a>Step 1: Install PerlBrew (Optional, only if your system perl is not usable)</h4>

<p>First of all download the perlbrew files and execute them:</p>

<div class="highlight highlight-source-shell"><pre>$ <span class="pl-cce">\c</span>url -L http://install.perlbrew.pl <span class="pl-k">|</span> bash</pre></div>

<p>Once it is installed follow the instructions to edit your .bash_profile and then install Perl using Perlbrew:</p>

<div class="highlight highlight-source-shell"><pre>$ perlbrew install perl-5.16.0</pre></div>

<p>This step will take some time to complete. Once it has you then need to install cpanm using perlbrew:</p>

<div class="highlight highlight-source-shell"><pre>perlbrew install-cpanm
cpanm --local-lib=<span class="pl-k">~</span>/perl5 <span class="pl-k">local</span>::lib <span class="pl-k">&amp;&amp;</span> <span class="pl-c1">eval</span> <span class="pl-s"><span class="pl-pds">$(</span>perl -I <span class="pl-k">~</span>/perl5/lib/perl5/ -Mlocal::lib<span class="pl-pds">)</span></span></pre></div>

<h4>
<a id="step-2-install-required-perl-packages" class="anchor" href="#step-2-install-required-perl-packages" aria-hidden="true"><span class="octicon octicon-link"></span></a>Step 2: Install Required Perl Packages</h4>

<p>There are number of perl libraries that are required, please install these from the command line as follows (alternatively you can use sudo cpan if your installation requires this):</p>

<div class="highlight highlight-source-shell"><pre>$ cpanm Text::xSV
$ cpanm SQL::Abstract
$ cpanm DBI
$ cpanm Data::Dumper
$ cpanm File::Path
$ cpanm File::Basename
$ cpanm Getopt::Std</pre></div>

<h4>
<a id="step-3-install-sqllite3" class="anchor" href="#step-3-install-sqllite3" aria-hidden="true"><span class="octicon octicon-link"></span></a>Step 3: Install SQLlite3</h4>

<p>ABBA uses SQLite in order to store and process locations. It can be installed as follows:</p>

<div class="highlight highlight-source-shell"><pre>sudo apt-get install sqlite3 libsqlite3-dev</pre></div>

<h4>
<a id="step-3-install-required-r-packages" class="anchor" href="#step-3-install-required-r-packages" aria-hidden="true"><span class="octicon octicon-link"></span></a>Step 3: Install required R packages</h4>

<p>ABBA uses R to perform much of the statisical analysis and plotting.  In order to that the following libraries are required (Remember that if you intend to use a none system version of R when running on a cluster you will also need to ensure that this libraries are installed for that version):</p>

<div class="highlight highlight-source-r"><pre>install.packages(<span class="pl-s"><span class="pl-pds">"</span>ggplot2<span class="pl-pds">"</span></span>)
install.packages(<span class="pl-s"><span class="pl-pds">"</span>ggthemes<span class="pl-pds">"</span></span>)
install.packages(<span class="pl-s"><span class="pl-pds">"</span>sp<span class="pl-pds">"</span></span>)
install.packages(<span class="pl-s"><span class="pl-pds">"</span>INLA<span class="pl-pds">"</span></span>, <span class="pl-v">repos</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>http://www.math.ntnu.no/inla/R/stable<span class="pl-pds">"</span></span>)</pre></div>

<h4>
<a id="step-4-download-software-and-the-test-data-and-annotation-library" class="anchor" href="#step-4-download-software-and-the-test-data-and-annotation-library" aria-hidden="true"><span class="octicon octicon-link"></span></a>Step 4: Download software and the test data and annotation library</h4>

<p>First of all to download the software clone it from our github account as follows:</p>

<div class="highlight highlight-source-shell"><pre>git clone https://github.com/SystemsGeneticsSG/ABBA.git
<span class="pl-c1">cd</span> ABBA</pre></div>

<p>We have provided a script that will fetch all of the required data and create the folders required. This can be executed as follows:</p>

<div class="highlight highlight-source-shell"><pre>perl setup.pl -af</pre></div>

<p>where -a is the flag to download the annotations database for the test and -f is the flag to download the test data.</p>

<h3>

<a id="running-abba-serially" class="anchor" href="#running-abba-serially" aria-hidden="true"><span class="octicon octicon-link"></span></a>Input Data Format</h3>
<p>The input data for ABBA is very simple but requires that you follow some strict formatting conditions. For each chromosome you will need a file containing the data. Each file will need to following the following naming convention chr1_both.in. This format gives the chromosome name (eg chr1) and the strand setup (both). The file needs to be tab delimied. Within the file the first four columns describe the locus eg. chr1 702 703 -. The following columns describe the number methylated reads and total reads in each replicate, For example if you have a 2 replicates in each sample then a line in the file might look like this:</p>

<div class="highlight highlight-source-shell"><pre>chr1 3417  3418  - 9 10 8 10 0 10 0 10</pre></div>

<p> You can see full examples by downloading the test data and paper examples using setup.pl as described <a href="#step-4-download-software-and-the-test-data-and-annotation-library">above</a>.


<a id="running-abba-serially" class="anchor" href="#running-abba-serially" aria-hidden="true"><span class="octicon octicon-link"></span></a>Running ABBA serially</h3>

<p>The simplest way to run ABBA is serially on a local machine. This will prepare the required files, execute the analysis and plot the results one after an other. There are a number of flags that need to be set for the script to run as follows:</p>

<div class="highlight highlight-source-shell"><pre>perl ABBA.pl -f input/test_set/ -n 2 -r 4 -p ABBAtest -a rn4</pre></div>

<ul>
<li>-f is the directory with the data to process in</li>
<li>-s is the window size between chunks that will be used to split the genome, ie here chunks will always be at 3000bp apart.</li>
<li>-m this the minimum number of CpGs that have to be in each chunk. If this is not reached when a gap of 3000bp occurs then the file will not be split.</li>
<li>-n this is the number of samples what are being analysed, this should always be 2</li>
<li>-r this is the number of replicates for each sample.</li>
<li>-t this is the number of reads to be accepted as a valid CpG site.</li>
<li>-c the is the number of reliable CpG sites required at a certain locus to used in the analysis.</li>
<li>-p is the name you are giving to the project</li>
<li>a is the species that the annotation will come from</li>
<li>-o is the directory to use as a tmp</li>
<li>-w is the window size to place around each DMR when plotting</li>
<li>-d is the average difference between the two samples for a DMR to considered.</li>
<li>-z is the number of standard deviations from the mean that the difference has to be for a DMR to be considered.</li>
<li>-y is the CpG density required within a CpG to be considered.</li>
<li>-e is the type of DMR to consider (density/length)</li>
</ul>

<p>At any stage you can open output/progress.html to view how the algorithm is proceeding.</p>

<h3>
<a id="running-abba-on-a-cluster" class="anchor" href="#running-abba-on-a-cluster" aria-hidden="true"><span class="octicon octicon-link"></span></a>Running ABBA on a cluster</h3>

<p>Running ABBA serially may take some time if you have a large dataset or you don't apply strict filtering criteria. In order to help with this ABBA has been implemented with the Sun Grid Engine queing system in mind. In order for this to work the executing is split into two stages; firstly the setup and parallel processing and secondly the post processing and plotting. We refer to these two sections as he qsub_setup and qsub_recover. The first can be executed as follows (with assumption that the version of R that will be used is stored in $HOME/R/3.2.2/bin/):</p>

<div class="highlight highlight-source-shell"><pre>qsub -pe smp 2 -N test_setup -o cluster/ABBAtest_setup.output -e cluster/ABBAtest_setup.error perl ABBA.pl -f input/test_set/ -n 2 -r 4 -p ABBAtest -a rn4 -i qsub_setup -j /home/gmsojlr/R/3.2.2/bin/</pre></div>

<p>This will start a single job that will prepare the data and then spawn extra jobs for each chromosome. You monitor the progress of this stage in two ways. Firstly using qstat in the normal way:</p>

<div class="highlight highlight-source-shell"><pre>qstat</pre></div>

<p>You can also extract the progress from the sqlite database that ABBA generates as it is processing as follows:</p>

<div class="highlight highlight-source-shell"><pre><span class="pl-c1">echo</span> <span class="pl-s"><span class="pl-pds">"</span>select stage,max(counter)*100 from file_ticker group by stage;<span class="pl-pds">"</span></span> <span class="pl-k">|</span> sqlite3 dbs/YOURPROJECTNAME.sqlite</pre></div>

<p>This will display the percentage of each chromosome that has been analysed. It is best not to do this too often as it can create lock errors on the sqlite database.</p>

<p>Once this has run and all of the individual chromosome jobs has completed then you can run the second stage as follows:</p>

<div class="highlight highlight-source-shell"><pre>qsub -pe smp 2 -N test_setup -o cluster/ABBAtest_setup.output -e cluster/ABBAtest_setup.error perl ABBA.pl -f input/test_set/ -n 2 -r 4 -p ABBAtest -a rn4 -i qsub_recover</pre></div>

<a id="running-abba-serially" class="anchor" href="#running-abba-serially" aria-hidden="true"><span class="octicon octicon-link"></span></a>Viewing the results</h3>

<p>Once ABBA is complete the results are automatically written to a html page that you can view as follows:</p>

<div class="highlight highlight-source-shell"><pre>xdg-open output/PROJECTNAME/PROJECTNAME_results.html</pre></div>

<p>If you have run the analysis on a cluster you can either scp the output folder to a convienient location or if you ssh -X username@yourserver.com you should also be able to view the webpage in a X-forwarded web browser using the same xdg-open command.</p>

<p>If you wish to use the dmrs elsewhere then a bed file is also created and located in output/PROJECTNAME/dmrs.bed.</p>

<h3>
<a id="troubleshooting" class="anchor" href="#troubleshooting" aria-hidden="true"><span class="octicon octicon-link"></span></a>Troubleshooting</h3>

<p>Sometimes when running in parrell R can have problems with shared libraries (eg  error while loading shared libraries: libicuuc.so.52: cannot open shared object file: No such file or directory). If this happens one workaround is to locally install a version of R into your home directory as follows:</p>

<div class="highlight highlight-source-shell"><pre>$ wget http://cran.rstudio.com/src/base/R-3/R-3.2.2.tar.gz
$ tar xvf R-3.2.2.tar.gz
$ <span class="pl-c1">cd</span> R-3.2.2
$ ./configure --prefix=<span class="pl-smi">$HOME</span>/R/3.2.2/
$ make <span class="pl-k">&amp;&amp;</span> make install</pre></div>

<p>You will then need to run ABBA using the -j option to specify where the R installation is and to install the required packages into this version of R.</p>

<p>During our own testing we sometimes found that certain CRAN mirrors would not contain the required packages. This results in ar error say that package X is not available for R v3.x.x. We found that one way to avoid this is to specifiy the repo using the following:</p>

<div class="highlight highlight-source-r"><pre>install.packages(<span class="pl-s"><span class="pl-pds">"</span>ggplot2<span class="pl-pds">"</span></span>, <span class="pl-v">repos</span><span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">"</span>http://cran.cnr.berkeley.edu<span class="pl-pds">"</span></span>)</pre></div>

<h3>
<a id="development" class="anchor" href="#development" aria-hidden="true"><span class="octicon octicon-link"></span></a>Development</h3>

<p>Want to contribute? Great!</p>

<p>First off please contains <a href="https://github.com/OwenRackham" class="user-mention">@OwenRackham</a> so you can be setup as a contributer to the package.</p>

<h3>
<a id="todos" class="anchor" href="#todos" aria-hidden="true"><span class="octicon octicon-link"></span></a>Todos</h3>

<ul>
<li>Write Tests</li>
<li>Javascript DMR viewer</li>
<li>GO enrichment pipeline</li>
<li>Project name folder so that overwritting doesnt occur</li>
<li>qsub_recover reteives the settings from the database rather than relying on the settings being the same, this requires a new table with each of the paramaters in it. This would be pretty useful anyway! this should also include the FDR cutoff</li>
<li>change folder structure so that it is easy to download all of the results from one project independently</li>
<li>recover files that have been processed properly so that failed ones can be rerun on larger nodes.</li>
<li>
</ul>

<h2>
<a id="license" class="anchor" href="#license" aria-hidden="true"><span class="octicon octicon-link"></span></a>License</h2>

<p>MIT</p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/SystemsGeneticsSG/ABBA">ABBA</a> is maintained by <a href="https://github.com/SystemsGeneticsSG">SystemsGeneticsSG</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

            <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-71843585-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
